### PR TITLE
Dev dataprotection microsoft.data protection 2022 03 31 preview

### PR DESCRIPTION
**Pull request by amchandn**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Review by TechTress98 — APPROVED** [review]

**Review by TechTress98 — CHANGES_REQUESTED** [review]
Need merged

**Review by TechTress98 — APPROVED** [review]

**Comment by amchandn** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {
> > ```
> 
> This seems to be an extension resource. How is it created? I don't see a PUT call.
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634). [](commit_id = [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634), deletion_comment = False)

We currently only intend to support to list (GET) on this resource via extension resource.
Creation of this extension resource in the backend is supported via a PUT on the non-extension proxy resource in our RP.

File: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/examples/BackupInstanceOperations/ListBackupInstancesExtensionRouting.json [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    "resourceType": "servers",
+    "resourceName": "viveksipgtest",
+    "api-version": "2022-03-31-preview"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
+          {
+            "id": "/subscriptions/04cf684a-d41f-4550-9f70-7708a3a2283b/resourceGroups/000pikumar/providers/Microsoft.DataProtection/backupVaults/PratikPrivatePreviewVault1/backupInstances/harshitbi2",
```
Comment (j5lim) [inline specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/examples/BackupInstanceOperations/ListBackupInstancesExtensionRouting.json:15]: >/subscriptions/04cf684a-d41f-4550-9f70-7708a3a2283b/resourceGroups/000pikumar/providers/Microsoft.DataProtection/backupVaults/PratikPrivatePreviewVault1/backupInstances/harshitbi2 [](http://example.com/codeflow?start=19&length=179)

The resource ID is invalid. The backupInstances resource should be an extension of the request scope (the sql server).
It should be
/subscriptions/04cf684a-d41f-4550-9f70-7708a3a2283b/resourceGroups/000pikumar/providers/Microsoft.DBforPostgreSQL/servers/viveksipgtest/providers/Microsoft.DataProtection/backupVaults/PratikPrivatePreviewVault1/backupInstances/harshitbi2

**Comment by j5lim** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {

Is this only the scope supported? i.e. can backupInstances be created for a nested resource like Microsoft.SQL/servers/databases? If so, the path should be more generic like '/{resourceId}/providers/Microsoft.DataProtection/backupInstances'

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in ab02656. [](commit_id = ab02656ddbaca941037dac982fd4305b78cdf634, deletion_comment = False)

**Comment by j5lim** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {

You should support GET for an individual resource as well.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in ab02656. [](commit_id = ab02656ddbaca941037dac982fd4305b78cdf634, deletion_comment = False)

**Comment by j5lim** [comment]
> > > ```
> > > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {
> > > ```
> > 
> > 
> > This seems to be an extension resource. How is it created? I don't see a PUT call.
> > Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634). [](commit_id = [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634), deletion_comment = False)
> 
> @jorgecotillo , We currently only intend to support to list (GET) on this resource via extension resource. Creation of this extension resource in the backend is supported via a different API -
> 
> https://github.com/amchandn/azure-rest-api-specs/blob/ab02656ddbaca941037dac982fd4305b78cdf634/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2022-03-01/dataprotection.json#L1067

The PUT in the link is for creating an extension resource for a resource group. That API cannot be used for this path, which scope is a resource not a RG. You will need to add PUT under this path as well.

**Comment by amchandn** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {
> > ```
> 
> Is this only the scope supported? i.e. can backupInstances be created for a nested resource like Microsoft.SQL/servers/databases? If so, the path should be more generic like '/{resourceId}/providers/Microsoft.DataProtection/backupInstances'
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634). [](commit_id = [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634), deletion_comment = False)

Thanks @j5lim , I've incorporated the changes as you suggested.

File: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/examples/BackupInstanceOperations/ListBackupInstancesExtensionRouting.json (hunk 1 above)
Comment (amchandn) [reply specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/examples/BackupInstanceOperations/ListBackupInstancesExtensionRouting.json]: thanks @j5lim for pointing it out, I've fixed the value of id in the example now

**Comment by amchandn** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {
> > ```
> 
> You should support GET for an individual resource as well.
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634). [](commit_id = [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634), deletion_comment = False)

@j5lim , we currently plan to support only list on this extension resource named 'backupInstances' so that we can author ARM policies over the root resource and take appropriate action depending on the response of the list 'backupInstances' proxy resource.
We will introduce support for PUT and GET on this extension resource in upcoming API versions.

**Comment by amchandn** [comment]
> > > > ```
> > > > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}/providers/Microsoft.DataProtection/backupInstances": {
> > > > ```
> > > 
> > > 
> > > This seems to be an extension resource. How is it created? I don't see a PUT call.
> > > Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/preview/2022-03-31-preview/dataprotection.json:1016 in [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634). [](commit_id = [ab02656](https://github.com/Azure/azure-rest-api-specs/commit/ab02656ddbaca941037dac982fd4305b78cdf634), deletion_comment = False)
> > 
> > 
> > @jorgecotillo , We currently only intend to support to list (GET) on this resource via extension resource. Creation of this extension resource in the backend is supported via a different API -
> > https://github.com/amchandn/azure-rest-api-specs/blob/ab02656ddbaca941037dac982fd4305b78cdf634/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2022-03-01/dataprotection.json#L1067
> 
> The PUT in the link is for creating an extension resource for a resource group. That API cannot be used for this path, which scope is a resource not a RG. You will need to add PUT under this path as well.

@j5lim , the resource which is referred in the link that I've shared is a proxy child resource over Microsoft.DataProtection/BackupVaults tracked resource. The creation of entities in our service happens when a PUT call is made for Microsoft.DataProtection/BackupVaults/{vaultName}/BackupInstances/{backupInstanceName}.
This creation creates a mapping in the service that allows entities to be returned in the list backupInstances API for the proxy resource as well.

The reason to do this is -
we currently plan to support only list on this extension resource named 'backupInstances' so that we can author ARM policies over the root resource and take appropriate action depending on the response of the list 'backupInstances' extension resource.
We will introduce support for PUT and individual GET on this extension resource in upcoming API versions.

**Review by j5lim — APPROVED** [review]
Approved from ARM side.

**Comment by jorgecotillo** [comment]
There are issues with your approach, RBAC being one of the most importa t ones, how do you ensure that a user doing a collection GET doesn't get access to resources he/she is not meant to?

Also, in our wiki we call out incomplete CRUD as a violation to our best practices,  proxy resources should at least model a collection GET and single GET.

**Comment by amchandn** [comment]
> {resourceId}



> There are issues with your approach, RBAC being one of the most importa t ones, how do you ensure that a user doing a collection GET doesn't get access to resources he/she is not meant to?
> 
> Also, in our wiki we call out incomplete CRUD as a violation to our best practices, proxy resources should at least model a collection GET and single GET.



> There are issues with your approach, RBAC being one of the most importa t ones, how do you ensure that a user doing a collection GET doesn't get access to resources he/she is not meant to?
> 
> Also, in our wiki we call out incomplete CRUD as a violation to our best practices, proxy resources should at least model a collection GET and single GET.

thanks @jorgecotillo for this feedback. We are discussing internally in our team how we can enforce proper RBAC in case of LIST on this extension resource.
We'll also discuss the feasibility of supporting single GET in this API version itself.

We have planned that when we support PUT for this resource in the future, we'll leverage LinkedAccessCheck to ensure that no RBAC violation occurs during PUT.
@j5lim / @jorgecotillo do let us know your thoughts on this.

**Comment by j5lim** [comment]
> 

The RP should always do RBAC check on the linked resources during PUT (e.g. via linked access check).
For the collection GET, the user will require Microsoft.DataProtection/backupInstances/read permission on the request scope, which is different than the RBAC for backupVaults/backupInstances as Jorge pointed out. So it will be the RP's responsibility to check if the user has Microsoft.DataProtection/backupVaults/backupInstances/read permission on the backupInstances being returned from the collection GET. You can't use ARM's linked access check for this, but you can do via checkaccess API (search 'checkaccess' in the ARM wiki for details).

**Comment by amchandn** [comment]
@msyyc / @Azure/swagger-specs-pull-request-approvers / @TechTress98 , can you please merge this PR now?

**Comment by msyyc** [comment]
after contact with go owner, will ignore the go breaking change

**Pull request merged**